File: _platform/ingress/src/forward.ts
import { request as httpRequest, type IncomingMessage, type ServerResponse } from "node:http";
import type { Duplex } from "node:stream";
import { HOP_HEADER } from "./cluster.js";
import type { Peer } from "./peers.js";

// Hands a request to the peer holding the tunnel, over the private network as plain HTTP, Host intact so it routes
// exactly like one from the internet.
// Streams, not buffers, in both directions and for upgrades, with no timeout; fly-replay is not used since it caps the
// body and can't cross hosts.
// `PeerUnreachable` is the failure the cluster acts on, by forgetting the holder.

export class PeerUnreachable extends Error {
    constructor(
        readonly peer: Peer,
        cause: Error,
    ) {
        super(`peer ${peer.host}:${peer.port} unreachable: ${cause.message}`, { cause });
        this.name = `PeerUnreachable`;
    }
}

// Node's connect-level failures: the peer is not there, not that it refused the request.
const CONNECT_ERRORS = new Set([`ECONNREFUSED`, `EHOSTUNREACH`, `ENETUNREACH`, `ENOTFOUND`, `ETIMEDOUT`, `ECONNRESET`, `EAI_AGAIN`]);

const isConnectError = (error: Error): boolean => `code` in error && CONNECT_ERRORS.has(String(error.code));

const asPeerError = (peer: Peer, error: Error): Error => (isConnectError(error) ? new PeerUnreachable(peer, error) : error);

// Appends the browser's own address, so the holding machine's logs show a browser, not a peer.
const forwardedFor = (request: IncomingMessage): string => {
    const existing = request.headers[`x-forwarded-for`];
    const address = request.socket.remoteAddress ?? ``;
    return existing === undefined ? address : `${Array.isArray(existing) ? existing.join(`, `) : existing}, ${address}`;
};

const forwardHeaders = (request: IncomingMessage): Record<string, string | string[]> => {
    const headers: Record<string, string | string[]> = {};
    for (const [name, value] of Object.entries(request.headers)) {
        if (value !== undefined) {
            headers[name] = value;
        }
    }
    headers[HOP_HEADER] = `1`;
    headers[`x-forwarded-for`] = forwardedFor(request);
    return headers;
};

const upstreamOptions = (peer: Peer, request: IncomingMessage) => ({
    host: peer.host,
    port: peer.port,
    method: request.method,
    path: request.url,
    headers: forwardHeaders(request),
    // Host is already forwarded in the headers; the peer routes by that, not this option.
    setHost: false,
});

export const forwardRequest = (peer: Peer, request: IncomingMessage, response: ServerResponse): Promise<void> =>
    new Promise<void>((resolve, reject) => {
        const upstream = httpRequest(upstreamOptions(peer, request), (answer) => {
            response.writeHead(answer.statusCode ?? 502, answer.statusMessage, answer.headers);
            answer.on(`error`, (error: Error) => reject(error));
            answer.pipe(response).on(`finish`, () => resolve());
        });
        upstream.on(`error`, (error: Error) => reject(asPeerError(peer, error)));
        // Browser going away mid-body ends the upstream stream too.
        response.on(`close`, () => upstream.destroy());
        request.pipe(upstream);
    });

// Peer's 101 response line and headers, written back to the browser's socket verbatim.
const upgradeHead = (answer: IncomingMessage): string => {
    const lines = [`HTTP/1.1 ${answer.statusCode ?? 101} ${answer.statusMessage ?? `Switching Protocols`}`];
    for (let index = 0; index < answer.rawHeaders.length; index += 2) {
        lines.push(`${answer.rawHeaders[index]}: ${answer.rawHeaders[index + 1]}`);
    }
    return `${lines.join(`\r\n`)}\r\n\r\n`;
};

// Forwards the upgrade request until the peer answers with 101, then splices the two sockets and reads nothing further.
// A different answer is relayed as a refusal on the raw socket instead of a hang; nothing is written before the peer
// answers.
export const forwardUpgrade = (peer: Peer, request: IncomingMessage, socket: Duplex, head: Buffer): Promise<void> =>
    new Promise<void>((resolve, reject) => {
        const upstream = httpRequest(upstreamOptions(peer, request));
        upstream.on(`upgrade`, (answer, upstreamSocket, upstreamHead) => {
            socket.write(upgradeHead(answer));
            if (upstreamHead.length > 0) {
                socket.write(upstreamHead);
            }
            if (head.length > 0) {
                upstreamSocket.write(head);
            }
            socket.pipe(upstreamSocket).pipe(socket);
            socket.on(`error`, () => upstreamSocket.destroy());
            upstreamSocket.on(`error`, () => socket.destroy());
            resolve();
        });
        upstream.on(`response`, (answer) => {
            // Peer refused the upgrade: relay its response as the refusal, then close.
            const chunks: Buffer[] = [];
            answer.on(`data`, (chunk: Buffer) => chunks.push(chunk));
            answer.on(`end`, () => {
                const body = Buffer.concat(chunks);
                socket.end(
                    `HTTP/1.1 ${answer.statusCode ?? 502} ${answer.statusMessage ?? `Bad Gateway`}\r\n` +
                        `Content-Type: ${answer.headers[`content-type`] ?? `text/plain; charset=utf-8`}\r\n` +
                        `Content-Length: ${body.length}\r\n` +
                        `Connection: close\r\n\r\n${body.toString(`utf8`)}`,
                );
                resolve();
            });
        });
        upstream.on(`error`, (error: Error) => reject(asPeerError(peer, error)));
        upstream.end();
    });

File: _platform/ingress/src/forward.rs
//! Hands a request to the peer holding its sandbox's tunnel, over the private network as plain HTTP/1.1 with its Host
//! intact and the hop marked, so the peer routes it as it would one from the internet and never hands it on again.
//! Streams both ways; an upgrade dials a connection of its own, since an upgraded one never returns to a pool, and is
//! spliced as every carrier's is (`relay::exchange`).

use std::net::IpAddr;

use http::header::{self, HeaderMap, HeaderValue};
use http::uri::PathAndQuery;
use http::{Request, Response, Version};
use hyper_util::client::legacy::Client;
use hyper_util::client::legacy::connect::HttpConnector;
use hyper_util::rt::{TokioExecutor, TokioIo};
use tokio::net::TcpStream;

use crate::body::{self, Body};
use crate::cluster::HOP_HEADER;
use crate::peers::Peer;

#[derive(Debug)]
pub enum ForwardError {
    /// The peer is not there: forget it as the holder.
    Unreachable(String),
    /// The peer was reached and the exchange failed; its belief stands until it says otherwise.
    Failed(String),
}

pub struct Forwarder {
    client: Client<HttpConnector, Body>,
}

impl Default for Forwarder {
    fn default() -> Self {
        let mut connector = HttpConnector::new();
        connector.set_nodelay(true);
        Self {
            client: Client::builder(TokioExecutor::new())
                .set_host(false)
                .build(connector),
        }
    }
}

impl Forwarder {
    pub async fn request(
        &self,
        peer: &Peer,
        request: Request<Body>,
        host: &str,
        browser: IpAddr,
    ) -> Result<Response<Body>, ForwardError> {
        let (mut parts, sent) = request.into_parts();
        let path = parts.uri.path_and_query().map_or("/", PathAndQuery::as_str);
        parts.uri = format!("http://{}{path}", peer.address(peer.port))
            .parse()
            .map_err(|error| ForwardError::Failed(format!("no usable peer address: {error}")))?;
        parts.version = Version::HTTP_11;
        mark(&mut parts.headers, host, browser);
        let answer = self
            .client
            .request(Request::from_parts(parts, sent))
            .await
            .map_err(|error| {
                if error.is_connect() {
                    ForwardError::Unreachable(error.to_string())
                } else {
                    ForwardError::Failed(error.to_string())
                }
            })?;
        Ok(answer.map(body::incoming))
    }

    /// Until the peer answers 101 nothing reaches the browser; then the two sides are spliced. Any other answer is
    /// relayed as the peer's refusal.
    pub async fn upgrade(
        &self,
        peer: &Peer,
        request: Request<Body>,
        host: &str,
        browser: IpAddr,
    ) -> Result<Response<Body>, ForwardError> {
        let stream = TcpStream::connect(peer.address(peer.port))
            .await
            .map_err(|error| ForwardError::Unreachable(error.to_string()))?;
        let _ = stream.set_nodelay(true);
        let (mut parts, sent) = request.into_parts();
        let path = parts
            .uri
            .path_and_query()
            .map_or("/", PathAndQuery::as_str)
            .to_owned();
        parts.uri = path
            .parse()
            .map_err(|error| ForwardError::Failed(format!("no usable path: {error}")))?;
        parts.version = Version::HTTP_11;
        relay::for_next_hop(&mut parts.headers, true);
        mark(&mut parts.headers, host, browser);
        relay::exchange(TokioIo::new(stream), Request::from_parts(parts, sent))
            .await
            .map_err(|error| ForwardError::Failed(error.to_string()))
    }
}

// The hop mark, the Host the peer routes by, and the browser's address appended so the holder's logs show a browser.
fn mark(headers: &mut HeaderMap, host: &str, browser: IpAddr) {
    if !headers.contains_key(header::HOST)
        && let Ok(host) = HeaderValue::from_str(host)
    {
        headers.insert(header::HOST, host);
    }
    headers.insert(HOP_HEADER, HeaderValue::from_static("1"));
    let forwarded = match headers
        .get("x-forwarded-for")
        .and_then(|value| value.to_str().ok())
    {
        Some(earlier) => format!("{earlier}, {browser}"),
        None => browser.to_string(),
    };
    if let Ok(forwarded) = HeaderValue::from_str(&forwarded) {
        headers.insert("x-forwarded-for", forwarded);
    }
}
